File: backend/job-application-service/src/main/java/com/jobportal/jobapp/controllers/JobApplicationController.java
package com.jobportal.jobapp.controllers;

import com.jobportal.jobapp.dto.ApplicationRequestDto;
import com.jobportal.jobapp.dto.ApplicationResponseDto;
import com.jobportal.jobapp.service.JobApplicationService;
import jakarta.validation.Valid;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/applications")
@Validated
public class JobApplicationController {

    private final JobApplicationService jobApplicationService;

    public JobApplicationController(JobApplicationService jobApplicationService) {
        this.jobApplicationService = jobApplicationService;
    }

    @PostMapping
    public ResponseEntity<ApplicationResponseDto> applyToJob(@Valid @RequestBody ApplicationRequestDto request) {
        return new ResponseEntity<>(jobApplicationService.applyToJob(request), HttpStatus.CREATED);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApplicationResponseDto> getApplicationById(@PathVariable Integer id) {
        return ResponseEntity.ok(jobApplicationService.getApplicationById(id));
    }

    @GetMapping("/candidate/{candidateId}")
    public ResponseEntity<List<ApplicationResponseDto>> getApplicationsByCandidate(@PathVariable Integer candidateId) {
        return ResponseEntity.ok(jobApplicationService.getApplicationsByCandidate(candidateId));
    }

    @GetMapping("/job/{jobId}")
    public ResponseEntity<List<ApplicationResponseDto>> getApplicationsByJob(@PathVariable Integer jobId) {
        return ResponseEntity.ok(jobApplicationService.getApplicationsByJob(jobId));
    }

    @PutMapping("/{id}/status/{statusId}")
    public ResponseEntity<ApplicationResponseDto> updateApplicationStatus(
            @PathVariable Integer id,
            @PathVariable Integer statusId) {
        return ResponseEntity.ok(jobApplicationService.updateApplicationStatus(id, statusId));
    }
}

File: backend-dotnet/Services/JobPortal.JobApplicationService/Controllers/JobApplicationController.cs
using Microsoft.AspNetCore.Mvc;
using JobPortal.JobApplications.DTOs;
using JobPortal.JobApplications.Interfaces;

namespace JobPortal.JobApplications.Controllers;

[ApiController]
[Route("api/applications")]
public class JobApplicationController : ControllerBase
{
    private readonly IJobApplicationService _jobApplicationService;

    public JobApplicationController(IJobApplicationService jobApplicationService)
    {
        _jobApplicationService = jobApplicationService;
    }

    [HttpPost]
    public async Task<IActionResult> ApplyToJob([FromBody] ApplicationRequestDto request)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }
        var response = await _jobApplicationService.ApplyToJobAsync(request);
        return StatusCode(StatusCodes.Status201Created, response);
    }

    [HttpGet("{id:int}")]
    public async Task<ActionResult<ApplicationResponseDto>> GetApplicationById([FromRoute] int id)
    {
        var response = await _jobApplicationService.GetApplicationByIdAsync(id);
        return Ok(response);
    }

    [HttpGet("candidate/{candidateId:int}")]
    public async Task<ActionResult<List<ApplicationResponseDto>>> GetApplicationsByCandidate([FromRoute] int candidateId)
    {
        var response = await _jobApplicationService.GetApplicationsByCandidateAsync(candidateId);
        return Ok(response);
    }

    [HttpGet("job/{jobId:int}")]
    public async Task<ActionResult<List<ApplicationResponseDto>>> GetApplicationsByJob([FromRoute] int jobId)
    {
        var response = await _jobApplicationService.GetApplicationsByJobAsync(jobId);
        return Ok(response);
    }

    [HttpPut("{id:int}/status/{statusId:int}")]
    public async Task<ActionResult<ApplicationResponseDto>> UpdateApplicationStatus(
        [FromRoute] int id,
        [FromRoute] int statusId)
    {
        var response = await _jobApplicationService.UpdateApplicationStatusAsync(id, statusId);
        return Ok(response);
    }
}
